Augmented Documentation Build

awholegunch opened the pull request:
Passes as Docs Develop Build: http://builds.cask.co/browse/CDAP-DDB49-17

Artifacts: 
- ZIP: http://builds.cask.co/browse/CDAP-DDB49-17/artifact/shared/Docs-ZIP/
- HTML: http://builds.cask.co/browse/CDAP-DDB49-17/artifact/shared/Docs-HTML/

This documentation build generates documentation following a new build structure of one zip: (`cdap-docs-4.0.0-SNAPSHOT-web.zip`)  for `docs.cask.co` for Amazon `S3`.

Once merged, this PR will require back-porting to currently active branches.

Commit:
work-in-progress

Commit:
Updates to script for setting headers in pages.

Commit:
Remove commented-out code. w-i-p.

Commit:
w-i-p

Commit:
Merge branch 'develop' into docs/4.0_augment_doc_build

Commit:
Cleanup.

wolf31o2 commented on cdap-docs/_common/common-build.sh:
```diff
@@ -523,8 +507,9 @@ function rewrite() {
   # or if $3 & $4=='', substitutes text in-place in file $1, using sed command $2
-  cd ${SCRIPT_PATH}
   local rewrite_source=${1}
+  local current_directory=$(pwd)
```
Use `pwd -P`

wolf31o2 replied:
Actually, just remove this line. 😄

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/_common/common-build.sh:
```diff
@@ -523,8 +507,9 @@ function rewrite() {
-  cd ${SCRIPT_PATH}
   local rewrite_source=${1}
+  local current_directory=$(pwd)
+  cd ${SCRIPT_PATH}
```
Since we're going to be coming back, use `pushd ${SCRIPT_PATH}` here.

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/_common/common-build.sh:
```diff
@@ -555,6 +540,7 @@ function rewrite() {
     echo "    ${sub_string} -> ${new_sub_string} "
     sed -e "s|${sub_string}|${new_sub_string}|g" ${rewrite_source} > ${rewrite_target}
   fi
+  cd ${current_directory}
```
Just `popd` here.

awholegunch replied:
Fixed. Thanks!

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -460,18 +490,74 @@ function copy_docs_inner_level() {
+  rm -rf ${TARGET_PATH}/base 
+  rm -rf ${TARGET_PATH}/current
+  rm -rf ${TARGET_PATH}/${PROJECT_VERSION}
+  rm -rf ${TARGET_PATH}/*.zip
```
You're already in `${TARGET_PATH}`, so just rm -rf these all on one line.

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -460,18 +490,74 @@ function copy_docs_inner_level() {
+  cp -r ${HTML} base/en
+
+  # Add a redirect index.html file
+  echo "${REDIRECT_EN_HTML}" > base/index.html
```
Any reason this isn't just a file copy with the HTML checked in as a regular file, rather than the contents being in a variable? Is the variable sourced from a file and I'm just not seeing it?

awholegunch replied:
No, just how things evolved...created a file and checked it in.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -460,18 +490,74 @@ function copy_docs_inner_level() {
+    rewrite ${SCRIPT_PATH}/${COMMON_SOURCE}/${HTACCESS} ${TARGET_PATH}/current/.${HTACCESS} "<version>" "current"
+    zip -qr ${ZIP_DIR_NAME}.zip current/.${HTACCESS}
+  fi
+  mv ${ZIP_DIR_NAME}.zip ${ZIP_DIR_NAME}-current.zip
```
There's no error checking in any of this. If any operation fails, it continues through the rest of the function. This function will only fail if the very last `rm` fails. This is used in our builds, so everything should fail fast, except for tests.

awholegunch replied:
Added.

wolf31o2 commented on cdap-docs/tools/doc-change.py:
```diff
@@ -0,0 +1,194 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+# Copyright © 2016 Cask Data, Inc.
```
This needs the correct license header.

awholegunch replied:
Fixed.

Commit:
Changes following review and comments:
- Consolidated removal statements
- Added error checking
- Removed redirect as a variable and replaced it with a checked-in file
- Replaced caching of current working directory with "popd" and "pushd" commands
- Added correct license header

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -388,6 +390,34 @@ function _build_docs() {
+  echo "--------------------------------------------------------"
+  echo
+  _package_docs build-web ${WEB} ${TRUE}
+  return $?
```
This return is redundant.

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -460,18 +490,105 @@ function copy_docs_inner_level() {
+  if [ "${errors}" != "0" ]; then
+      echo "Could not cleanup old directories and zips"
+      return ${errors}   
+  fi
```
If you're using `-f` with `rm`, it'll always return 0, even on failure. There's no need to check the exit on this.

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -460,18 +490,105 @@ function copy_docs_inner_level() {
+  if [ "${errors}" != "0" ]; then
+      echo "Could not create base"
+      return ${errors}   
+  fi
```
Do numerical comparisons, here. `if [[ ${errors} -ne 0 ]]; then`

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -460,18 +490,105 @@ function copy_docs_inner_level() {
+  echo
+  echo "1: 'current' version, with canonical ref link ${ZIP_DIR_NAME}-current"
+  cp -r base current
+  python ${doc_change_py} ${TARGET_PATH} --current=${PROJECT_VERSION} current
```
This is the most likely place to fail and there's no error checking.

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -460,18 +490,105 @@ function copy_docs_inner_level() {
+  if [ "x${1}" == "x" ]; then
+    ZIP_DIR_NAME="${PROJECT}-docs-${PROJECT_VERSION}"
+  else
+    ZIP_DIR_NAME="${PROJECT}-docs-${PROJECT_VERSION}-$1"
```
Use `${zip_target}` here. Also, don't use string comparison to determine if a variable is set...
`if [[ -n ${zip_target} ]]; then`

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -460,18 +490,105 @@ function copy_docs_inner_level() {
+  fi
+  # Cleanup
+  rm -rf base current ${ZIP_DIR_NAME}-current.zip ${ZIP_DIR_NAME}-future.zip
+  return $?
```
This will always return 0, no matter what.

```
$ rm -rf /some/non-existent/directory ; echo $?
0
```

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/tools/doc-change.py:
```diff
@@ -0,0 +1,206 @@
+#
+# ./doc-change.py ../../cdap-gh-pages/cdap --ghpages --current=3.6.0 3.6.0
+#
+# ./doc-change.py ../../cdap-gh-pages/cdap --ghpages --current=3.6.0 current
```
Why? What should I expect to see? There's no point in documenting this if you don't put the expected results.

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/tools/doc-change.py:
```diff
@@ -0,0 +1,206 @@
+    """
+
+    parser = OptionParser(
+        usage="%prog [options] doc-directory [doc-set]",
```
You show "prog options dir set" here, but use "prog dir options set" in all of your examples in the comments. Be consistent.

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/tools/doc-change.py:
```diff
@@ -0,0 +1,206 @@
+    for root, dirs, files in os.walk(doc_set_path):
+        for f in files:
+            if f.endswith('.html'):
+                html_count += 1
```
I think that you should assign the matching files to an array here, which you will loop through below. There's no reason to reset `html_count` and reuse it as you're performing the same conditional check below.

awholegunch replied:
Good idea. Fixed.

wolf31o2 commented on cdap-docs/tools/doc-change.py:
```diff
@@ -0,0 +1,206 @@
+                sys.stdout.flush()
+    print "\nConverted HTML files: %s" % html_count
+
+def convert_upper_level_html(html_path):
```
This method is used once. Any reason we can't just do it in-line below?

awholegunch replied:
Better than that: you found a bug. That method isn't doing anything. Fixed.

Commit:
Edits per comments.

awholegunch commented:
Edited and re-running build: http://builds.cask.co/browse/CDAP-DDB49-6

awholegunch commented:
"On-hold" as the last run, though it passed, did not adjust the docs correctly. The edits may have broken something...

Commit:
Fix processing and packaging of docs.

awholegunch commented:
Re-running build: http://builds.cask.co/browse/CDAP-DDB49-7

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -342,17 +345,16 @@ function stash_github_zip() {
   echo
-  mv ${TARGET_TEMP}/*.zip ${TARGET_PATH}
-  rm -rf ${TARGET_TEMP}
+  mv ${TARGET_TEMP}/*archive.zip ${TARGET_PATH}
```
If this fails, the function returns 0 due to the final echo.

awholegunch replied:
Fixed. Similar issue in function before this.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -388,6 +388,43 @@ function _build_docs() {
+  echo "Packaging Web Docs"
+  echo "--------------------------------------------------------"
+  echo
+  _package_docs build-web ${WEB} ${TRUE}
```
`${BUILD_WEB}`

awholegunch replied:
fixed.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -388,6 +388,43 @@ function _build_docs() {
+  echo "Packaging GitHub Docs"
+  echo "--------------------------------------------------------"
+  echo
+  _package_docs build-github ${GITHUB} ${FALSE}
```
`${BUILD_GITHUB}`

awholegunch replied:
fixed.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -460,18 +497,128 @@ function copy_docs_inner_level() {
+    zip_dir_name="${PROJECT}-docs-${PROJECT_VERSION}"
+  else
+    zip_dir_name="${PROJECT}-docs-${PROJECT_VERSION}-$1"
+  fi
```
So the logic is `if $2 then append $1`?

wolf31o2 replied:
Also:

```
local zip_dir_name="${PROJECT}-docs-${PROJECT_VERSION}"
[[ -n ${zip_target} ]] && zip_dir_name += "-${doc_target}"
```

awholegunch replied:
Actually, this should be using `${zip_target}` :

```
  local zip_dir_name="${PROJECT}-docs-${PROJECT_VERSION}"
  [[ -n ${zip_target} ]] && zip_dir_name += "-${zip_target}"
```

wolf31o2 replied:
Yeah, that's what I thought, but I was writing based on the original logic.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -460,18 +497,128 @@ function copy_docs_inner_level() {
+  fi
+  local zip_dir_name_archive
+  local build_flag
+  if [ "x${doc_target}" == "x${BUILD_GITHUB}" ]; then
```
`if [[ ${doc_target} == ${BUILD_GITHUB} ]]; then`

Don't use x$something in bash expressions.

awholegunch replied:
fixed.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -460,18 +497,128 @@ function copy_docs_inner_level() {
+  
+  echo "Creating base"
+  mkdir base
+  cp -r ${HTML} base/en
```
Chain these: `mkdir base && cp -r ${HTML} base/en` and your error checking will catch either failure.

awholegunch replied:
fixed.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -460,18 +497,128 @@ function copy_docs_inner_level() {
+      return ${errors}   
+  fi
+  zip -qr ${zip_dir_name}.zip current/* --exclude *.DS_Store* *.buildinfo*
+  if [[ "x${zip_extras}" != "x" ]]; then
```
`if [[ -n ${zip_extras} ]]; then`

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -460,18 +497,128 @@ function copy_docs_inner_level() {
+  if [[ ${errors} -ne 0 ]]; then
+      echo "Could not move ${zip_dir_name}.zip to ${zip_dir_name}-current.zip"
+      return ${errors}   
+  fi
```
What errors are you expecting to happen here? We're sure that `${zip_dir_name}.zip` exists due to previous error checking. If we always want to overwrite `${zip_dir_name}-current.zip`, then just use `mv -f` and drop the error checking. Is there another case I'm missing?

wolf31o2 replied:
Even better, why don't we just create `${zip_dir_name}-current.zip` and not bother with the `mv`? Seems these three sections could be refactored into a single function and have less maintenance in the long run, as it appears to be 3 slight variations on the same steps.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -460,18 +497,128 @@ function copy_docs_inner_level() {
+  if [[ ${errors} -ne 0 ]]; then
+      echo "Could not move ${PROJECT_VERSION} to future"
+      return ${errors}   
+  fi
```
OK, but why? I don't see the contents of `future` being used after this.

awholegunch replied:
Because then you (I) can look at the directory results and check that the contents are correct, either locally on on the Bamboo build. True, you should really download the zips and unpack them (which I do as a final check) but while I am getting this going, it allows for an intermediate check that things are ok.

wolf31o2 replied:
Testing/debug code doesn't belong in a "final" code review/pull request, though. It's just a convenience and not something used by the build. I think it should be removed.

wolf31o2 commented on cdap-docs/tools/doc-change.py:
```diff
@@ -0,0 +1,205 @@
+# 
+# - If the canonical numbered version:
+#   no changes
+#   If basename is not the numbered version to be used for empty title tags, use
```
As in, bash `basename`? Some other meaning?

awholegunch replied:
The Python method `basename` is used to extract the last component of the path; that's the basename or the doc set name. But we don't necessarily want to use that (it might be "current") for fixing titles that are incorrect; sometimes we want to use the version instead.

wolf31o2 replied:
So more like "if the version derived from the basename of $foo is not the numbered version"...

awholegunch replied:
Yes. Will change.

Commit:
Fixes per comments.

Commit:
Clean up per comments.

awholegunch commented:
Running a new build: http://builds.cask.co/browse/CDAP-DDB49-8

Commit:
Major cleanup. Added new google tag manager code to template.

awholegunch commented:
Passes a new Quick Build: http://builds.cask.co/browse/CDAP-DDB49-9.

Commit:
Add an alias to docs-set of docs-all

awholegunch commented:
Running a new Build: http://builds.cask.co/browse/CDAP-DDB49-10

Commit:
Update name of deployment ZIP file, as there is now only a single version.

Commit:
Correct the name of the zip file.

awholegunch commented:
Passes as a new Build: http://builds.cask.co/browse/CDAP-DDB49-12

HTML: http://builds.cask.co/browse/CDAP-DDB49-12/artifact/shared/Docs-HTML/

wolf31o2 commented on cdap-docs/_common/common-build.sh:
```diff
@@ -71,6 +56,7 @@ if [ "x${2}" == "x" ]; then
 else
   PROJECT_PATH="${SCRIPT_PATH}/../../../${2}"
 fi
+PROJECT_PATH=`python -c "from os import path; print(path.realpath('${PROJECT_PATH}'));"`
```
Why is this added? Is it necessary to be in a canonical PATH? If so, can we avoid shelling out to python?
`PROJECT_PATH=$(cd ${PROJECT_PATH}; pwd -P)`

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/_common/common-build.sh:
```diff
@@ -224,36 +153,17 @@ function build_extras() {
 
 function check_includes() {
-  if [ "x${DOCS_LOCAL}" == "x${TRUE}" ]; then
+  if [ "${DOCS_LOCAL}" == "${TRUE}" ]; then
```
Use `[[` and get rid of all the quotes, let bash handle it

awholegunch replied:
Done. And in `build.sh`.

wolf31o2 commented on cdap-docs/_common/common-build.sh:
```diff
@@ -337,10 +247,10 @@ function download_file() {
     local target=${includes_dir}/${target_filename}
+  else
+    local target=${includes_dir}/${file_name}
   fi
```
FYI: this whole thing can be done in bash without conditionals...
`local target=${includes_dir}/${target_filename:-${file_name}}`

I'll leave it up to you if you'd like to change these, but generally this sort of comparison makes more sense to do in pure shell, as `[` is often an external symlink to the external `test` command versus the shell built-in...

awholegunch replied:
OK...

wolf31o2 commented on cdap-docs/_common/common-build.sh:
```diff
@@ -354,10 +264,48 @@ function download_file() {
+  mkdir ${SCRIPT_PATH}/${LICENSES_PDF}
+  # paths are relative to location of $DOC_GEN_PY script
+  LIC_PDF="../../../${REFERENCE}/${LICENSES_PDF}"
+  LIC_RST="../${REFERENCE}/source/${LICENSES}"
```
Does it make more sense to make these absolute PATHs? This allows us to move the `$DOC_GEN_PY` script and not have to find and edit these relative PATHs, that would otherwise be difficult to correlate with the script move.

awholegunch replied:
It does, though that might mean changing the `doc-gen.py` script.

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -20,283 +20,210 @@
+# NO_JAVADOCS: Set to not build Javadocs, no matter which actions are used (for testing)
 
 source ./vars
 source _common/common-build.sh
```
This requires that this script is executed as `./build.sh` and it'll fail if invoked any other way. `cd $(cd $(dirname ${BASH_SOURCE[0]}); pwd -P)` will ensure your working directory is the canonical location for the script, no matter how it's called (unless it's called from a symlink outside this location, but we're not concerned about that use case)...

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -20,283 +20,210 @@
+  build_javadocs docs
+  build_docs_cli
+  build_docs_second_pass
+  build_docs_package
```
What if any of these fail?

awholegunch replied:
I don't consider it a huge problem; I would rather run through them all, and get all the errors in a single pass, rather than having to keep running over and over, fixing the new erros that surface. Typically one section will have errors, but other parts will pass.

wolf31o2 replied:
OK, but this code isn't written in a way for that to make sense. A failure in one of these functions will cause cascading failures/problems. They're not written as fully independent units of work. Therefore, the script should fail fast, rather than continue generating false-positive errors (along with actual errors, yes) and using up build resources.

awholegunch replied:
I'm not sure about that: any of those four can fail without it cascading to the next. The Javadocs and the CLI docs can be (and are) ignored if they fail or aren't built, and if there is a failure in the second_pass, you can still package what was copied. 

I looked over these functions, and I'm not sure what errors I would promulgate, or what codes I would pass that would be meaningful. Many of the functions in these functions don't return error codes themselves, as they don't really have an impact outside themeselves. We have a system that captures Sphinx and build errors and puts them in a file that gets read and reported on at the end, and I think that is working well enough. 

I'm having trouble seeing the benefit to spending a lot of time here.

wolf31o2 replied:
I'm reading that it's acceptable to package incomplete/partial builds, versus failing. If that's the case, I'm going to just let this go. It doesn't make any sense to me, and you haven't been able to explain why failure is acceptable. To me, the end result is bogus. If files that are supposed to be copied, aren't, then how is downstream supposed to know that something's gone wrong? If this were in some kind of test suite, that would be one thing. However, this is a build. Single failures should be enough to be fatal and it should fail fast. This is a complete anti-pattern and not something to be encouraged, IMO.

At any rate, it's not worth fighting over. Feel free to merge without fixing this.

awholegunch replied:
I see your point. Maybe we need it both ways: when this is run as build+deploy (or as a build on Bamboo) it needs to behave as you describe, but locally (on a Mac) allow that behaviour be over-ridden.

wolf31o2 replied:
Not for this PR, but later, you could implement a "test_or_die" function or something, which checks if tests should be enabled, and dies if not. This would allow both sets of functionality. 😄

awholegunch replied:
As we need to get this done, I will merge with this behavour, but add an issue to correct this: https://issues.cask.co/browse/CDAP-7573

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -20,283 +20,210 @@
+    set_mvn_environment
+    echo "JAVA_HOME: ${JAVA_HOME}"
+    echo "JAVA Version:"
+    java -version
```
Only if the `java` in the PATH is the same one in `JAVA_HOME`... maybe set a `JAVA` variable with the resolved PATH to the correct `java` binary, like [CDAP does](https://github.com/caskdata/cdap/blob/develop/cdap-common/bin/functions.sh#L286-L310)... or, you could actually source the `functions.sh` file and use the function, directly.

awholegunch replied:
Fixed. Sourced `functions.sh`.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -20,283 +20,210 @@
+    mvn -version
+    local start=`date`
+    echo "Maven clean install"
+    MAVEN_OPTS="-Xmx1024m -XX:MaxPermSize=128m" mvn clean install -P examples,templates,release -DskipTests -Dgpg.skip=true
```
WHOA!!! This is doing a `mvn` install on our build hosts? Is that necessary? This will replace the most-recent built JARs in the local Maven cache with whatever's in your branch, even though it's not been unit tested. This is a major red flag.

Can you replace this and the `javadoc_run` below, along with the conditional above?

```
if [[ ${javadoc_type} == ${ALL} ]]; then
  javadoc_run="javadoc:aggregate"
else
  javadoc_run="site"
fi
MAVEN_OPTS="-Xmx4g -XX:MaxPermSize=256m" # match other CDAP builds
mvn clean package ${javadoc_run} -P examples,templates,release -DskipTests -Dgpg.skip=true -DisOffline=false ${debug_flag}
```

awholegunch replied:
AFAIK, this is required to build the Javadocs correctly. I will find Terence and ask him to look at this, as he helped put this together.

awholegunch replied:
Terence suggests:

"[that to] avoid affect the current user local repo cache, you can provide the `-Dmaven.repo.local=/some/path` option; the downside is, of course, it will take the build longer to run… as each time it will fetch the dependencies from remote repo."

awholegunch replied:
I think the two runs are required, and when we attempted to put them together it failed. The thought was this was a consequence of the dependencies in CDAP, that two runs were required; the product had to be built first so that the different modules could link together correctly. Sort of like the docs with a multi-manual project: two runs are required there to enable the index to be correctly linked.

wolf31o2 replied:
Can we verify, and if so, comment somewhere? I'd rather know than assume and using a second maven repository would work, but not preferable for obvious reasons. This can be a separate task from this merge, since this code was there previously. I just saw it and realized how it could impact certain builds using the agent after a Docs build.

awholegunch replied:
But I will test what you wrote...and see if what we have can be improved.

awholegunch replied:
OK, I implemented what you wrote. Especially since what we currently have doesn't work for `javadocs-all` under Java 8.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -395,20 +246,20 @@ function build_docs_inner_level() {
     echo
     cd $SCRIPT_PATH/${i}
-    ./build.sh ${1} ${ARG_2}
+    ./build.sh ${1}
```
Since all of the `build.sh` arguments appear to map 1:1 to function calls, why don't use just call the functions, rather than having the script call itself with different arguments? That's really... dodgy.

awholegunch replied:
It's not calling itself: it is calling the `build.sh` in each manual, each of which is customized for the manual. They all share the `common-build.sh`. 

The idea is that the same command can be used at both the individual manual level and at the doc-set level: `build.sh docs`. It makes the edit-build-check cycle simple.

The top level script changes into each manual, and runs the same command there, but it is running different code.

wolf31o2 replied:
Gotcha. Maybe a comment to make it more clear to the reader.

awholegunch replied:
Added comments at start of the function.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -499,47 +370,48 @@ function clean_outer_level() {
+  PROJECT_PATH="${SCRIPT_PATH}/.."
+  PROJECT_PATH=$(cd ${PROJECT_PATH} && pwd)
+  SOURCE_PATH="${SCRIPT_PATH}/../../"
+  SOURCE_PATH=$(cd ${SOURCE_PATH} && pwd)
```
`pwd -P` will give you the physical location

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -499,47 +370,48 @@ function clean_outer_level() {
+    echo "Did not find MANUAL \"${MANUAL}\": are you in the correct directory?"
+    exit ${E_WRONG_DIRECTORY}
+  fi
+  exit 1
```
Functions should never exit (except main), IMO. They should always "return" and it should be up to the caller to determine whether to exit or not.

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/tools/doc-change.py:
```diff
@@ -0,0 +1,162 @@
+# Changes the "current" doc set; sets empty titles, adds canonical link to 3.6.0
+# 
+# python doc-change.py --type=robots ~/Source/cdap/cdap-docs/target/3.6.0
+# Changes the "3.6.0" doc set; sets empty titles, adds a robots no-index meta-tag
```
Doesn't this also set canonical links?

awholegunch replied:
No, it's one or the other: cannonical links or robots.txt.

wolf31o2 replied:
OK, then the `--type=canonical` appears to be redundant.

awholegunch replied:
OK, changed to `--robots`, and a true-false flag instead.

wolf31o2 reviewed (changes requested):
Few changes

Commit:
Fixes per comments.

Commit:
Rename Python script to match pattern.
Remove redirects by properly describing paths.
Fix name of docs-change script.

wolf31o2 commented on cdap-docs/_common/common-build.sh:
```diff
@@ -288,11 +288,11 @@ function copy_license_pdfs() {
+  if [[ ${OSTYPE} == "darwin"* ]]; then
     # TODO: hard-coded Java version 1.7
     export JAVA_HOME=$(/usr/libexec/java_home -v 1.7)
+    java -version
```
This will still always return whichever Java version is returned first in your PATH, not necessarily the version in `JAVA_HOME`... if you use `${JAVA}` from `functions.sh`'s `cdap_set_java`, then it'll always be the version in `JAVA_HOME`. Since I don't handle the Mac case, you'll want to still do that and set JAVA_HOME on Mac before running the `cdap_set_java` function, followed by using `${JAVA} -version` here.

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -169,27 +170,34 @@ function build_javadocs() {
+    set_mvn_environment
+    source ${PROJECT_PATH}/cdap-common/bin/functions.sh
+    cdap_set_java
     java -version
```
`${JAVA} -version` otherwise, you're getting what's in the `PATH`, not `JAVA_HOME`.

Maven will respect JAVA_HOME.

awholegunch replied:
Fixed.

Commit:
Merge branch 'develop' into docs/4.0_augment_doc_build

Commit:
Reset the zip name to include web so current deployment doesn't break.

Commit:
Fixes the passing of titles. Removes unused variables.
Uses ${JAVA} consistently when calling Java.
Other cleanup.

Commit:
Move setting of cdap_set_java. Always show Java version.

wolf31o2 commented:
👍 LGTM
